rmangal blogpost (#525)
ropensci/roweb2: content/blog/2019-10-21-rmangal.md

---
slug: rmangal
title: 'rmangal: making ecological networks easily accessible'
date: 2019-10-21
authors:
  - Kevin Cazelles
  - Steve Vissault
categories: blog

> topicid: 1848

Reviewer: ```suggestion
topicid: 
```

I'll fill topicid right before publishing

Author: :+1:

tags:
  - Software Peer Review
  - R
  - packages
  - open-data
  - data
  - data-access
  - ecology
  - ecological networks
  - API
  - API client
  - Mangal
  - rmangal
---

In early September, the version 2.0.0 of [rmangal][rmangal] was approved by

> rOpenSci, four weeks latter it made the CRAN. Following-up on our experience we

Reviewer: ```suggestion
rOpenSci, four weeks later it made it to CRAN. Following-up on our experience we
```
typo & English style

detail below the reasons why we wrote rmangal, why we submitted our package to

> rOpenSci and how the peer-review improve our package.

Reviewer: ```suggestion
rOpenSci and how the peer review improved our package.
```
English style
no hyphen in peer review

> ### [Mangal](https://mangal.io/#/), a database for ecological networks

Reviewer: Best not to link Mangal in heading since it doesn't look nice

Author: :+1:

Reviewer 2: Done

Ecological networks are defined as a set of species populations (the nodes of
the network) connected through ecological interactions (the edges). Interactions
are ecological processes in which one species affects another. Although
predation is probably the most known and documented interaction, other less
noticeable associations are just as essential to ecosystem functioning. For

> instance, a mammal that unintentionally disperses viable seeds attached to his

Reviewer: ```suggestion
instance, a mammal that unintentionally disperses viable seeds attached to its
```
his -> its

Author: Ooopsy! :+1:

fur might help plants to thrive. All of these interactions occur simultaneously,

> shaping ecosystems functioning and making them as complex as they are

Reviewer: in "ecosystems functioning" in two places, should there be an apostrophe? "ecosystems' functioning"

Author: **ecosystem functioning** is better I think.

Reviewer 2: Changes made in bc4ec8312afa16c90730371cdf250d6bb7203884

fascinating.

> Recording and storing properly these interactions help ecologists to better

Reviewer: ```suggestion
Recording and properly storing these interactions help ecologists to better
```
English flow

understand ecosystems. That is why they are currently compiling datasets to
explore how species associations vary over environmental gradients and how
species lost might affect ecosystems functioning. This fundamental research
question should help us understanding how ecological networks will respond to
global change. To this end, the Mangal project <https://mangal.io/#/>
standardizes ecological networks and eases their access. Every dataset contains
a collection of networks described in a specific reference (a scientific
publication, a book, etc.). For every network included in the database, Mangal

> include all the species name and several taxonomic identifiers

Reviewer: ```suggestion
includes all the species names and several taxonomic identifiers
```

([gbif](https://www.gbif.org/), [eol](https://eol.org/),

> [tsn](https://www.itis.gov/) etc.) as well as all interaction and their type.

Reviewer: ```suggestion
[tsn](https://www.itis.gov/) etc.) as well as all interactions and their types.
```

Currently, Mangal includes 172 datasets, which represents over 1300 ecological
networks distributed worldwide.

### An R client to make ecological networks easily accessible

In 2016, the first paper describing the project was published[^original_ref]. In

> 2018, a substantial effort was put in order to improve the data structure and

Reviewer: ```suggestion
2018, a substantial effort was made in order to improve the data structure and
```

gather new networks from existing publications. In 2019, the web API was
rewritten, a new website launched and hundreds of new interactions were added.

> {{< figure src="/img/blog-images/2019-10-22-rmangal/mangal_explorer.png"

Reviewer: This image is great in that it gives a glimpse of what the tool gives a user. However, it made me crave the ability to zoom in and explore. Any way you can link to that specific example in the site https://mangal.io/#/network? Would help readers quickly get a deeper understanding.

At the end of the section "Mangal, a database for ecological networks" I got very curious about having a specific example, so maybe there's an opportunity to write ~2 lines about an example from one dataset and link to that network in the site. That might be powerful.

Reviewer: sounds good

Reviewer 2: Fix in ab660256e07bf53e9726ed34fc3128e708d4f516

Reviewer 2: We enhanced the post with a quick example, see ae9added2f15ba713e5305c3b07fe1815535ce87

 alt="write an alternative text describing the information contained in the image" width="100%" >}}

Because of all these modifications, the [first version of

> rmangal](https://github.com/mangal-wg/rmangal-v1), the R client for the web API,

Reviewer: don't link to old version, unless you have a really strong reason to do that

Author: noted!

Reviewer 2: This has been removed in bc4ec8312afa16c90730371cdf250d6bb7203884

was obsolete and a new version needed. It is worth explaining here why the R
client is an important component of the Mangal project. Even though Mangal has a
documented [RESTful API](https://mangal.io/doc/api/), this web technology is not
commonly used by ecologists. On the contrary, providing a R client ensures that

> the scientific community that document these interactions on the field can

Reviewer: ```suggestion
the scientific community that documents these interactions in the field can
```

access them, as easily as possible. The same argument holds true for the [Julia
client](https://mangal.io/doc/jl/) that [Timothée Poisot](https://poisotlab.io/)
wrote because [Julia](https://julialang.org/) is increasingly popular among
theoreticians, that can test ecological theory with such datasets.

> We had two main objectifs for [rmangal 2.0.0][rmangal]. First, the rmangal

Reviewer: ```suggestion
We had two main objectives for [rmangal 2.0.0][rmangal]. First, the rmangal
```
objectifs -> objectives

package had to allow users to search for all entries in the database in a very
flexible way. From a technical point this means that we had to write functions
to query all the endpoints of the [new web API][API]. The second goal was to
make the package as user friendly as possible. To do so, we used explicit and
consistent names for functions and arguments, designed a simple workflow, and
documented how to use other field related packages (such as [igraph][igraph]) to
visualize and analyze networks. All of this is explain in the vignette ["get

> started with rmangal"](ttps://ropensci.github.io/rmangal/articles/rmangal.html).

Reviewer: ```suggestion
started with rmangal"](https://ropensci.github.io/rmangal/articles/rmangal.html).
```
fixed broken link

> ### A successful peer-review process

Reviewer: ```suggestion
### A successful peer review process
```
no hyphen in peer review throughout post

<!-- version to prerelease -->

> After some hard work behind the screen and once we deemed our two objectifs

Reviewer: ```suggestion
After some hard work behind the screen and once we deemed our two objectives
```

> achieved, we decided to submit rmangal package to rOpenSci for peer-review. We

Reviewer: ```suggestion
achieved, we decided to submit the rmangal package to rOpenSci for peer review. We
```

did so because we needed feedback, we needed qualified people to critically

> assess whether our two main objectifs were achieved. Given the strong expertise

Reviewer: ```suggestion
assess whether our two main objectives were achieved. Given the strong expertise
```

our work and pointed out areas where improvement was required. One good example
of how their review made our package better concerns the dependencies. We
originally listed [sf][sf] in `Imports` as we used it to filter networks based
on geographic coordinates. But the reviewers pointed out that this was not an
essential part of the package and that [sf][sf] has several dependencies. This  made us realize that for one extra feature, we were substantially
increasing the number of indirect dependencies. Following the reviewers'
suggestions, we moved [sf][sf] to `Suggests` and added a message to warn users
that to use the [spatial filtering
feature](https://ropensci.github.io/rmangal/articles/rmangal.html#geolocalize-mangal-networks-with-sf)
requires sf to be installed. Similarly, based on another good comment, we
added a function to convert Mangal networks into [tidygraph][tidygraph] objects
and we documented how to plot Mangal networks with [ggraph][ggraph] (and so we
added those packages in `Suggests`). Such improvements were very helpful to
properly connect rmangal to the existing R packages. The plethora of R packages

> is one of its major strength, and connecting a package properly to other makes

Reviewer: ```suggestion
is one of its major strengths, and connecting a package properly to others makes
```

the entire ecosystem even stronger.

<!-- Future -->

> We are now looking for user feedback, not only for rmangal but also for the web

Reviewer: Good here to link to pkg, API and site one last time in this sentence

Author: :+1:  I would ad something here.

Reviewer 2: Added in 3487f18

> API and the website. We welcome suggestions and contributions, especially for

Reviewer: Do you have any specific areas in which you'd like feedback? How do you prefer people contact you for this e.g. open an issue?

This whole post reads well, so I don't want to mess with it, but giving some specifics here if possible, might make the difference in people actually taking a step to contribute.

Author: Hum, I think it is a good point we were thinking about posting issue for feedback, I'll add something!

Author: :+1:

Reviewer 2: Changes in 3487f1825a3ed7876e77e84a4c8a0421c048da02

the documentation. In the future, we envision that rmangal will integrate
functions to format ecological networks for ecologists willing to add their
datasets to Mangal. This will likely be the next major release of rmangal.

### Acknowledgments

We are thankful to all [contributors to
rmangal](https://ropensci.github.io/rmangal/authors.html) and to all ecologists
that have spent countless hours in collecting data. We would like to thank [Anna
Willoughby][arw36] and [Thomas Lin Pedersen][thomasp85] for thorough reviews as
well as [Noam Ross](https://github.com/noamross) for handling the review
process.

> [^original_ref]: Poisot, T. et al. mangal - making ecological network analysis simple. Ecography 39, 384–390 (2016).

Reviewer: ```suggestion
[^original_ref]: Poisot, T. et al. mangal - making ecological network analysis simple. Ecography 39, 384–390 (2016). https://doi.org/10.1111/ecog.00976
```
link to paper

Author: :+1:

[arw36]: https://github.com/arw36
[thomasp85]: https://github.com/thomasp85
[igraph]: https://igraph.org/r/
[tidygraph]: https://tidygraph.data-imaginist.com/
[ggraph]:https://ggraph.data-imaginist.com/

> [rmangal]: https://ropensci.github.io/rmangal/

Reviewer: Consider linking to https://docs.ropensci.org/rmangal/ (automatically built and hosted documentation) and adding that url to https://github.com/ropensci/rmangal. This is 

About: https://ropensci.org/technotes/2019/06/07/ropensci-docs/

Author: https://docs.ropensci.org/rmangal/ added on the repo! Do you want to change it in the post?

Reviewer: Can you please make the changes in the post yourself, and then once you've finished I can make minor changes that arise later if needed

Author: Will do :+1: !

[RESTful API](https://mangal.io/doc/api/), this web technology is not commonly
used by ecologists. On the contrary, providing a R client ensures that the
scientific community that documents these interactions in the field can access
them, as easily as possible. The same argument holds true for the [Julia
client](https://mangal.io/doc/jl/) that [Timothée Poisot](https://poisotlab.io/)
wrote because [Julia](https://julialang.org/) is increasingly popular among
theoreticians, that can test ecological theory with such datasets.

We had two main objectives for [rmangal 2.0.0][rmangal]. First, the rmangal
package had to allow users to search for all entries in the database in a very
flexible way. From a technical point this means that we had to write functions
to query all the endpoints of the [new web API][API]. The second goal was to
make the package as user friendly as possible. To do so, we used explicit and
consistent names for functions and arguments. We then designed a simple workflow, and
documented how to use other field related packages (such as [igraph][igraph]) to
visualize and analyze networks (see below). You can find further details in the vignette ["get

> started with rmangal"](https://ropensci.github.io/rmangal/articles/rmangal.html).

Reviewer: ```suggestion
started with rmangal"](https://docs.ropensci.org/rmangal/articles/rmangal.html).
```
updated other link to https://docs.ropensci.org/rmangal/articles/rmangal.html

achieved, we decided to submit the rmangal package to rOpenSci for peer review. We
did so because we needed feedback, we needed qualified people to critically
assess whether our two main objectives were achieved. Given the strong expertise
of rOpenSci in software review, and given that our package was in-scope,
submitting rmangal to rOpenSci was an obvious choice.

We [had very valuable
feedback](https://github.com/ropensci/software-review/issues/332) from [Anna
Willoughby][arw36] and [Thomas Lin Pedersen][thomasp85]. They carefully assessed
our work and pointed out areas where improvement was required. One good example
of how their review made our package better concerns the dependencies. We
originally listed [sf][sf] in `Imports` as we used it to filter networks based
on geographic coordinates. But the reviewers pointed out that this was not an
essential part of the package and that [sf][sf] has several dependencies. This  made us realize that for one extra feature, we were substantially
increasing the number of indirect dependencies. Following the reviewers'
suggestions, we moved [sf][sf] to `Suggests` and added a message to warn users
that to use the [spatial filtering

> feature](https://ropensci.github.io/rmangal/articles/rmangal.html#geolocalize-mangal-networks-with-sf)

Reviewer: ```suggestion
feature](https://docs.ropensci.org/rmangal/articles/rmangal.html#geolocalize-mangal-networks-with-sf)
```

requires sf to be installed. Similarly, based on another good comment, we
added a function to convert Mangal networks into [tidygraph][tidygraph] objects
and we documented how to plot Mangal networks with [ggraph][ggraph] (and so we
added those packages in `Suggests`). Such improvements were very helpful to
properly connect rmangal to the existing R packages. The plethora of R packages
is one of its major strengths, and connecting a package properly to others makes
the entire ecosystem even stronger.

<!-- Future -->
We are now looking for user experience feedback, not only for rmangal
([vignette](https://docs.ropensci.org/rmangal/articles/rmangal.html)) but also
for the web API ([documentation](https://mangal.io/doc/api/)) and the [mangal.io
website](https://mangal.io). We welcome suggestions and contributions,

> especially for the documentation by opening new issues on Github

Reviewer: ```suggestion
especially for the documentation by opening new issues on GitHub
```
